tests/store/tracking/test_file_store.py

```diff
@@ -1130,6 +1139,17 @@ def test_log_batch(self):
         )
         self._verify_logged(fs, run_id, metric_entities, param_entities, tag_entities)
 
+    def test_log_batch_max_length_value(self):
+        param_entities = [Param("long param", list(range(500))), Param("short param", "xyz")]
+        expected_param_entities = [
+            Param("long param", str(list(range(500)))),
```

Comment:
@johnyNJ Are you sure this line is correct?

```python
>>> str(list(range(500)))
'[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221, 222, 223, 224, 225, 226, 227, 228, 229, 230, 231, 232, 233, 234, 235, 236, 237, 238, 239, 240, 241, 242, 243, 244, 245, 246, 247, 248, 249, 250, 251, 252, 253, 254, 255, 256, 257, 258, 259, 260, 261, 262, 263, 264, 265, 266, 267, 268, 269, 270, 271, 272, 273, 274, 275, 276, 277, 278, 279, 280, 281, 282, 283, 284, 285, 286, 287, 288, 289, 290, 291, 292, 293, 294, 295, 296, 297, 298, 299, 300, 301, 302, 303, 304, 305, 306, 307, 308, 309, 310, 311, 312, 313, 314, 315, 316, 317, 318, 319, 320, 321, 322, 323, 324, 325, 326, 327, 328, 329, 330, 331, 332, 333, 334, 335, 336, 337, 338, 339, 340, 341, 342, 343, 344, 345, 346, 347, 348, 349, 350, 351, 352, 353, 354, 355, 356, 357, 358, 359, 360, 361, 362, 363, 364, 365, 366, 367, 368, 369, 370, 371, 372, 373, 374, 375, 376, 377, 378, 379, 380, 381, 382, 383, 384, 385, 386, 387, 388, 389, 390, 391, 392, 393, 394, 395, 396, 397, 398, 399, 400, 401, 402, 403, 404, 405, 406, 407, 408, 409, 410, 411, 412, 413, 414, 415, 416, 417, 418, 419, 420, 421, 422, 423, 424, 425, 426, 427, 428, 429, 430, 431, 432, 433, 434, 435, 436, 437, 438, 439, 440, 441, 442, 443, 444, 445, 446, 447, 448, 449, 450, 451, 452, 453, 454, 455, 456, 457, 458, 459, 460, 461, 462, 463, 464, 465, 466, 467, 468, 469, 470, 471, 472, 473, 474, 475, 476, 477, 478, 479, 480, 481, 482, 483, 484, 485, 486, 487, 488, 489, 490, 491, 492, 493, 494, 495, 496, 497, 498, 499]'
>>> len(str(list(range(500))))
2390
```

Reply:
validation works on `param.value` which in this case is within limit.  After stringification, the length can be much longer. I wonder if this is as intended..

Reply:
It looks like we don't stringify a param value at all:

https://github.com/mlflow/mlflow/blob/ca5b76c1cb3823b06c26e464f466d585703ebd7d/mlflow/entities/param.py#L7-L30

I think we need to:

1. Stringify the value in the constructor.
2. Stringify the value in the `value` method.

```diff
diff --git a/mlflow/entities/param.py b/mlflow/entities/param.py
index 157bc5882..2b9cc7e52 100644
--- a/mlflow/entities/param.py
+++ b/mlflow/entities/param.py
@@ -17,7 +17,7 @@ class Param(_MLflowObject):
                 key = key.name
                 value = str(value)
         self._key = key
-        self._value = value
+        self._value = str(value)  # 1
 
     @property
     def key(self):
@@ -27,7 +27,7 @@ class Param(_MLflowObject):
     @property
     def value(self):
         """String value of the parameter."""
-        return self._value
+        return str(self._value)  # 2
 
     def to_proto(self):
         param = ProtoParam()
```

@dbczumar wdyt?

Reply:
@harupy  I don't think so, we do stringify it in client 
[here](https://github.com/mlflow/mlflow/blob/master/mlflow/tracking/_tracking_service/client.py)


```python
def log_param(self, run_id, key, value):
        """
        Log a parameter (e.g. model hyperparameter) against the run ID. Value is converted to
        a string.
        """
        param = Param(key, str(value))
        try:
            self.store.log_param(run_id, param)
        except MlflowException as e:
            if e.error_code == ErrorCode.Name(INVALID_PARAMETER_VALUE):
                msg = f"{e.message}{PARAM_VALIDATION_MSG}'"
                raise MlflowException(msg, INVALID_PARAMETER_VALUE)
            else:
                raise e
```

So the question is: do we want to apply limit 500 on value **after** stringification or **before**?
Current logic of both master and my PR both checks the limit before so in case a user passes non-string value, the stringfication will really blow up the value and it will still  be stored.

Reply:
- We should check the length after stringification.
- We should not pass a non-string value to `Param`.

Reply:
and to complicate further, `log_batch` which is called by `log_params` does not seem to stringify at all which will cause `SQLAlchemyStore` to fail when doing something like `log_params({'p':[1.0,2.0]})`

Reply:
> log_batch which is called by log_params does not seem to stringify at all

We stringify the param values in `log_params`:

https://github.com/mlflow/mlflow/blob/0fa849ad75e5733bf76cc14a4455657c5c32f107/mlflow/utils/autologging_utils/client.py#L183

https://github.com/mlflow/mlflow/blob/45e543b9c9a0828b21dd37bc7d032f93f717c545/mlflow/tracking/fluent.py#L696

Reply:
I was partially mistaken. The stringification happens upon `Param` construction and either store will never store anything longer than 500 regardless of whether original value was stringified or not.